### PR TITLE
Created ControlValues for controlled gates/operations, fix for #4512

### LINKED ISSUE #4512: Controlled gate construction cannot construct some cases
**Is your feature request related to a use case or problem? Please describe.**

Suppose you want a three qubit controlled gate like `(|00><00| + |11><11|) X + (|01><01| + |10><10| I` which flips the third qubit if the first qubit is either 00 or 11.  This currently isn't possible to construct given the way cirq specified `control_values`.    This is because `control_values` is a sequence of length equal to the number of controls, and this sequence is either `int`s or sequences of `int`s. In the second case the sequence means "either of these".  Thus you might think this would construct the gate above, but it does not
```
cirq.X(q2).controlled_by(q0, q1, control_values=((0, 1), (0, 1))
```
(this flips the third qubit for all inputs)
Or maybe this
```
cirq.X(q2).controlled_by(q0, q1, control_values=((0, 0), (1, 1))
```
(this flips the third qubit if the first control is zero and the second is one.

**Describe the solution you'd like**

Not sure we can fix this now, as this interface is used throughout :(

Probably the better interface would have been sequence of ints, same interperation, sequence of sequence either zipped up to get possible control positive values, or sequence of sequence where inner sequence is sequence of ints describing a positive control value.  If you really want to include unions for any of these, places where there is an int could have either an indicator value or another sequence to unionize what this means.

**[optional] Describe alternatives/workarounds you've considered**

Create two ops in a row that do what you want.  Custom gate controlled gate.

**What is the urgency from your perspective for this issue? Is it blocking important work?**

P2 - we should do it in the next couple of quarters

### PR DESCRIPTION
**Pull request by NoureldinYosri**
- created control_values.py which contains the ControlValues class.
- FreeVars and ConstrainedVars classes are provided for ease of use.
- while the basic idea of ControlValues integrating it inside the code base was challening
- the old way of using control_values assumed it's a tuple of tuples of ints and was used as thus (comparasion, hashing, slicing, fomatting, conditioning, and loops), the ControlValues class had to provide these functionalities
- the trickiest part to get right was the support for formatting!
- fixes #4512

### CODE REVIEW

**Comment by tanujkhattar** [comment]
Thanks @NoureldinYosri for opening up the PR. Unit tests need to be added for every new code / modification or else the coverage check won't pass and we can't approve any PRs with failing tests. Please add the unit tests (and fix other failing tests) and then we can do a review for the PR. 

You can look at https://github.com/quantumlib/Cirq/blob/master/docs/dev/development.md#continuous-integration-and-local-testing to learn more about how to run the integration tests locally.

**Comment by NoureldinYosri** [comment]
@tanujkhattar sorry for spamming your email with the submissions, I have been running the tests locally but I also had a bash script to do the pull request which was buggy :3, there is also an issue two of the checks are contradictory the Changed files test and Type check want different things in the type hint, I want to use `: ControlValues` but the change file test gives `E   NameError: name 'ControlValues' is not defined` one of them accepts only `Type[ControlValues]`, the other accepts only `Type['ControlValues']` -_-

**Comment by NoureldinYosri** [comment]
@tanujkhattar can you check this PR?, all checks pass now :)

File: cirq-core/cirq/ops/control_values.py
Comment (dabacon) [inline cirq-core/cirq/ops/control_values.py:78]: I assume this is camel case to avoid conflict with `and`?  Suggest using `product`.  That would match the pythonic way to do a cartesian product, `itertools.product`

File: cirq-core/cirq/ops/control_values_test.py
Comment (dabacon) [inline cirq-core/cirq/ops/control_values_test.py:31]: we have a helper for testing equality.  could you and an EqualsTester test?

File: cirq-core/cirq/ops/control_values.py
Comment (dabacon) [inline cirq-core/cirq/ops/control_values.py:177]: should this be a f-string?

File: cirq-core/cirq/ops/control_values_test.py
Comment (dabacon) [inline cirq-core/cirq/ops/control_values_test.py:84]: you can get rid of the pylint warning by doing
`_ = control_vals[0:1]`

File: cirq-core/cirq/ops/control_values.py
Comment (dabacon) [inline cirq-core/cirq/ops/control_values.py:137]: I guess the intention here is to make this equal to other possible ways to build a control value?  If so I think you want to check the types more and if it doesn't equal any of those types, return `NotImplemented` (python's equals contract)

File: cirq-core/cirq/ops/controlled_gate_test.py
Comment (dabacon) [inline cirq-core/cirq/ops/controlled_gate_test.py:558]: are we testing the new forms here?  I would add a test for these new cases.

**Review by dabacon — COMMENTED** [review]
Some initial comments.

File: cirq-core/cirq/ops/control_values.py
Comment (NoureldinYosri) [reply cirq-core/cirq/ops/control_values.py]: done

File: cirq-core/cirq/ops/control_values.py
Comment (NoureldinYosri) [reply cirq-core/cirq/ops/control_values.py]: done

File: cirq-core/cirq/ops/control_values.py
Comment (NoureldinYosri) [reply cirq-core/cirq/ops/control_values.py:23]: no, flatten works recusively while itertools.chain.from_iterable yields from only the second level, context https://docs.python.org/3/library/itertools.html#itertools.chain.from_iterable

File: cirq-core/cirq/ops/control_values_test.py
Comment (NoureldinYosri) [reply cirq-core/cirq/ops/control_values_test.py:35]: done

File: cirq-core/cirq/ops/control_values_test.py
Comment (NoureldinYosri) [reply cirq-core/cirq/ops/control_values_test.py]: done

File: cirq-core/cirq/ops/control_values.py
Comment (NoureldinYosri) [reply cirq-core/cirq/ops/control_values.py:140]: done

File: cirq-core/cirq/ops/controlled_gate_test.py
Comment (NoureldinYosri) [reply cirq-core/cirq/ops/controlled_gate_test.py:558]: I assume you are talking about mixed gates, no?
can you suggest a test(s)?

File: cirq-core/cirq/ops/control_values.py
Comment (NoureldinYosri) [reply cirq-core/cirq/ops/control_values.py]: done

**Comment by NoureldinYosri** [comment]
@daxfohl @dabacon this PR has been waiting for sometime, can we get it reviewed :)

File: cirq-core/cirq/ops/control_values.py
Comment (daxfohl) [inline cirq-core/cirq/ops/control_values.py:51]: Members should be private, with getter properties when necessary.

File: cirq-core/cirq/ops/control_values.py
Comment (daxfohl) [inline cirq-core/cirq/ops/control_values.py:108]: Can this be done without the copy module?

**Comment by daxfohl** [comment]
Can you add "fixes #4512" to the description so they are linked?

File: cirq-core/cirq/ops/controlled_gate_test.py
Comment (daxfohl) [reply cirq-core/cirq/ops/controlled_gate_test.py:558]: I think he means that there are no tests that pass the new class into the ControlledGate constructor and verify the functionality works as desired. There should be tests that do this, preferably with conditions that were not previously expressible.

File: cirq-core/cirq/ops/control_values.py
Comment (daxfohl) [inline cirq-core/cirq/ops/control_values.py:56]: Does this need to be a member? Frequently classes that iterate via member iterators have problems when iterating multiple times, especially with multiple iterations going on at once. It's usually better to create the iterators independently in the methods that do the iteration.

**Review by daxfohl — COMMENTED** [review]
One more thing

**Comment by NoureldinYosri** [comment]
I addressed most of the comments, I have a couple of questions which I put in the comments.

File: cirq-core/cirq/ops/control_values.py
Comment (daxfohl) [reply cirq-core/cirq/ops/control_values.py:46]: The root `__init__.py` file, so users can do `cirq.ControlValues`. (Once you do this you'll find it triggers some additional tests that check for serializability etc. These tasks need to be completed too, or else the control won't be able to run on real machines).

File: cirq-core/cirq/ops/control_values.py
Comment (daxfohl) [reply cirq-core/cirq/ops/control_values.py]: If that's the case, you shouldn't need *any* copy since tuples are immutable. If you remove the copy completely does something break?

File: cirq-core/cirq/ops/control_values.py
Comment (daxfohl) [inline cirq-core/cirq/ops/control_values.py:79]: All public methods need type annotations

File: cirq-core/cirq/ops/control_values.py
Comment (daxfohl) [inline cirq-core/cirq/ops/control_values.py:91]: This needs to be implemented in a functional fashion. Cirq gates are supposed to be immutable (this has been overlooked in places, but there's an effort to clean those up). This modifies the `ControlValues` object in-place, which given it's a member of a gate, provides a backdoor to mutating the gate.

File: cirq-core/cirq/ops/control_values.py
Comment (daxfohl) [inline cirq-core/cirq/ops/control_values.py:249]: Again, needs to be implemented functionally, not in-place modification.

File: cirq-core/cirq/ops/control_values.py
Comment (daxfohl) [inline cirq-core/cirq/ops/control_values.py:104]: Once everything has been implemented functionally, there should be no need for this function.

**Review by daxfohl — CHANGES_REQUESTED** [review]

**Comment by daxfohl** [comment]
@NoureldinYosri consider using `@dataclass(frozen=True)` on this to help ensure immutability. We use that elsewhere in cirq, you can just search for examples. It provides nice default implementation of equality, hash, etc as well. Just an option--if it gets in the way then don't worry about it.

File: cirq-core/cirq/ops/control_values.py
Comment (NoureldinYosri) [reply cirq-core/cirq/ops/control_values.py]: done

File: cirq-core/cirq/ops/control_values.py
Comment (NoureldinYosri) [reply cirq-core/cirq/ops/control_values.py]: done

File: cirq-core/cirq/ops/control_values.py
Comment (NoureldinYosri) [reply cirq-core/cirq/ops/control_values.py]: added

**Comment by NoureldinYosri** [comment]
@daxfohl @dabacon @tanujkhattar ptal

**Comment by daxfohl** [comment]
The changes resolve my concerns about mutability. Still need to add these classes to the root namespace in `cirq-core/cirq/__init__.py`. Also the classes' top level docstrings need some more detailed explanation and examples suitable for end users. Beyond that, I think I'll pass this along to the maintainers at this point.

**Comment by MichaelBroughton** [comment]
There are some concerns here re: qsim compatability @95-martin-orion , this has implications for TFQ as well and our goals for decompose always returning X,Y,Z,C*Z. Have you looked into this Orion ?

**Review by 95-martin-orion — CHANGES_REQUESTED** [review]
Quick review pass + follow up to Mike's question

File: cirq-core/cirq/ops/controlled_operation.py
Comment (NoureldinYosri) [reply cirq-core/cirq/ops/controlled_operation.py:145]: the `|00><00| + |11><11|` combination can be added in the same way that the xor combination `|01><01| + |10><10|` which was not supported in the old way and can be seen in the tests

File: cirq-core/cirq/ops/control_values.py
Comment (NoureldinYosri) [reply cirq-core/cirq/ops/control_values.py:29]: this assumes that lists of combinations are the only way that wil be used to define control_values, there was a previous comment on this PR about possibly supporting other ways like tensors or sympy expressions in the future.

the linked-list structure allows defining using different ways at the same time e.g. CV1(a table) -> CV2(a sympy expression) -> ... etc

each of those object will inheret from the ControlValues base class so that wil work perfectly.

**Review by NoureldinYosri — COMMENTED** [review]
I'll remove the debug line and add the classes to the root namespace ASAP.

the main point of the link-list structure is all for different ways of creating  control_values (e.g. the old way, the new way, using tensors, using generators, using sympy expressions ..etc) to be used together.

File: cirq-core/cirq/ops/control_values.py
Comment (daxfohl) [reply cirq-core/cirq/ops/control_values.py:29]: The comment referred to above is on the linked issue. https://github.com/quantumlib/Cirq/issues/4512#issuecomment-1014129817

**Comment by 95-martin-orion** [comment]
Bump: the checks appear to have gotten stuck 😖 . @NoureldinYosri, you may need to push a change (maybe resolving the conflicts marked) to get it moving again.

File: cirq-core/cirq/ops/controlled_operation.py
Comment (daxfohl) [reply cirq-core/cirq/ops/controlled_operation.py:145]: Did we ever have a design review on this one? I provided the initial code reviews assuming a design was approved, but I still kind of stand by what I originally said on the linked issue, that this could be done more simply by just adding a bool flag to the ControlledGate constructor. That flag would control whether to interpret the 2D input array vertically as allowed states by qubit (which is how it currently works, but is limited), or horizontally as a full set of allowed states (which allows full control). 

That this question came up during code review makes me think end users are going to have trouble understanding how to use these classes.

Comment (NoureldinYosri) [reply cirq-core/cirq/ops/controlled_operation.py:145]: I introduced the idea in https://github.com/quantumlib/Cirq/issues/4512#issuecomment-952284810 and talked about it ina cirq meeting https://github.com/quantumlib/Cirq/issues/4512#issuecomment-953208347, wrote a prototype in https://github.com/quantumlib/Cirq/issues/4512#issuecomment-954142591 and got a review in https://github.com/quantumlib/Cirq/issues/4512#issuecomment-974322956

--------------------------------------------------------------
the boolean idea has some issues
1. it doesn't address your previous comment  https://github.com/quantumlib/Cirq/issues/4512#issuecomment-1014129817, it fact it will make it harder to add those new ways lalter.
2. there is no way of getting around computing products, what a boolean will do is to force the users to compute them before passing them to the constructor, a work around would be passing an array of booleans but that has its own issues.
3. any work around will make the constructors for controlledGate and controlledOperation execture complex logic which is not a good idea in general, introduce tech debt and make it harder change or support new types later,

-----------------------------------------------------------------
lets discuss this in the next cirq meeting.

Comment (daxfohl) [reply cirq-core/cirq/ops/controlled_operation.py:145]: Sounds good. I have limited availability for cynqs these days, but I think a reprise with the rest of the team on Wed would be great just to get the ball rolling on this PR, since it seems to keep going stagnant. I'm looking forward to the functionality.

File: cirq-core/cirq/ops/control_values.py
Comment (daxfohl) [reply cirq-core/cirq/ops/control_values.py:29]: Another thought, have you considered making a separate `ControlValuesProduct` class that merges two (or maybe N) ControlValues objects into a product? That approach would allow for a smaller base class and adhere more to the single responsibility pattern.

File: cirq-core/cirq/ops/control_values.py
Comment (daxfohl) [inline cirq-core/cirq/ops/control_values.py:188]: Let's bikeshed this during synq. I don't think users will find the name intuitive. Perhaps `PerQubitControlValues`? Also the docstring could use some examples to make it clear.

**Review by daxfohl — COMMENTED** [review]
Added a couple of thoughts that we could go over during cynq.

File: cirq-core/cirq/ops/controlled_gate.py
Comment (daxfohl) [inline cirq-core/cirq/ops/controlled_gate.py:263]: Does this still round-trip? Given there's no type information about which subclass of `ControlValues` it uses, I'm not seeing how it could.

**Review by daxfohl — COMMENTED** [review]
One more question

**Comment by NoureldinYosri** [comment]
@daxfohl @95-martin-orion I have a simple way of implementing this which I think addresses all the comments/concerns, I will present it during the cirq meeting.

**Comment by daxfohl** [comment]
Great! One other thing to look out for is the circuit diagram. The current
logic labels each qubit with the control values. That probably won't work
for the new scenarios. Will need a different approach for them.

**Comment by 95-martin-orion** [comment]
Discussed on Cirq sync: @NoureldinYosri will write a document capturing the overarching design of this change, we'll discuss and reach a consensus there, then any changes can be reflected here.

**Comment by NoureldinYosri** [comment]
As discussed in the cirq-sync meeting I wrote a [design](https://docs.google.com/document/d/1TtHd4Cfyo9Zwrxtols5N_Ffv_dbXH2XgBB6D3tNIUYo/edit?usp=sharing&resourcekey=0-Zwlj47ZlUM626InJWsvWiA) descriping the current approach.

@tanujkhattar @95-martin-orion @daxfohl @dabacon

**Comment by tanujkhattar** [comment]
This can now be closed after https://github.com/quantumlib/Cirq/pull/5788

**Pull request closed without merging**
